File: paimon-core/src/test/java/org/apache/paimon/table/system/BucketsTableTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.table.system;

import org.apache.paimon.data.BinaryString;
import org.apache.paimon.data.GenericRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.data.serializer.InternalRowSerializer;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.fs.Path;
import org.apache.paimon.fs.local.LocalFileIO;
import org.apache.paimon.predicate.Predicate;
import org.apache.paimon.predicate.PredicateBuilder;
import org.apache.paimon.reader.RecordReader;
import org.apache.paimon.schema.Schema;
import org.apache.paimon.schema.SchemaManager;
import org.apache.paimon.schema.SchemaUtils;
import org.apache.paimon.schema.TableSchema;
import org.apache.paimon.table.FileStoreTable;
import org.apache.paimon.table.FileStoreTableFactory;
import org.apache.paimon.table.Table;
import org.apache.paimon.table.TableTestBase;
import org.apache.paimon.table.source.ReadBuilder;
import org.apache.paimon.types.DataTypes;
import org.apache.paimon.utils.Projection;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.apache.paimon.catalog.Identifier.SYSTEM_TABLE_SPLITTER;
import static org.apache.paimon.table.system.BucketsTable.BUCKETS;
import static org.assertj.core.api.Assertions.assertThat;

/** Unit tests for {@link BucketsTable}. */
public class BucketsTableTest extends TableTestBase {

    private Table bucketsTable;

    @BeforeEach
    public void before() throws Exception {
        String tableName = "MyTable";
        FileIO fileIO = LocalFileIO.create();
        Path tablePath = new Path(String.format("%s/%s.db/%s", warehouse, database, tableName));
        Schema schema =
                Schema.newBuilder()
                        .column("pk", DataTypes.INT())
                        .column("pt", DataTypes.INT())
                        .column("col1", DataTypes.INT())
                        .partitionKeys("pt")
                        .primaryKey("pk", "pt")
                        .option("bucket", "2")
                        .build();
        TableSchema tableSchema =
                SchemaUtils.forceCommit(new SchemaManager(fileIO, tablePath), schema);
        FileStoreTable table =
                FileStoreTableFactory.create(LocalFileIO.create(), tablePath, tableSchema);

        bucketsTable = catalog.getTable(identifier(tableName + SYSTEM_TABLE_SPLITTER + BUCKETS));

        write(table, GenericRow.of(1, 1, 1), GenericRow.of(1, 2, 5));
        write(table, GenericRow.of(1, 1, 3), GenericRow.of(1, 2, 4));
    }

    @Test
    public void testBucketsTable() throws Exception {
        assertThat(read(bucketsTable, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{1}"), 0, 2L, 2L),
                        GenericRow.of(BinaryString.fromString("{2}"), 0, 2L, 2L));
    }

    @Test
    public void testBucketsTableWithPartitionFilter() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(BucketsTable.TABLE_TYPE);

        Predicate filter = builder.equal(0, BinaryString.fromString("{1}"));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{1}"), 0, 2L, 2L));

        filter = builder.equal(0, BinaryString.fromString("{2}"));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{2}"), 0, 2L, 2L));

        filter =
                builder.in(
                        0,
                        Arrays.asList(
                                BinaryString.fromString("{1}"), BinaryString.fromString("{2}")));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{1}"), 0, 2L, 2L),
                        GenericRow.of(BinaryString.fromString("{2}"), 0, 2L, 2L));

        filter = builder.equal(0, BinaryString.fromString("{999}"));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4})).isEmpty();
    }

    @Test
    public void testBucketsTableWithBucketFilter() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(BucketsTable.TABLE_TYPE);

        Predicate filter = builder.equal(1, 0);
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{1}"), 0, 2L, 2L),
                        GenericRow.of(BinaryString.fromString("{2}"), 0, 2L, 2L));

        filter = builder.equal(1, 1);
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4})).isEmpty();
    }

    @Test
    public void testBucketsTableWithCombinedFilter() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(BucketsTable.TABLE_TYPE);

        Predicate filter =
                PredicateBuilder.and(
                        builder.equal(0, BinaryString.fromString("{1}")), builder.equal(1, 0));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4}))
                .containsExactlyInAnyOrder(
                        GenericRow.of(BinaryString.fromString("{1}"), 0, 2L, 2L));

        filter =
                PredicateBuilder.and(
                        builder.equal(0, BinaryString.fromString("{1}")), builder.equal(1, 1));
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4})).isEmpty();
    }

    @Test
    public void testBucketsTableUnsupportedPredicateFallsBackToFullScan() throws Exception {
        PredicateBuilder builder = new PredicateBuilder(BucketsTable.TABLE_TYPE);

        // isNotNull cannot be pushed down as partition filter — must fall back to full scan
        Predicate filter = builder.isNotNull(0);
        assertThat(readWithFilter(bucketsTable, filter, new int[] {0, 1, 2, 4})).hasSize(2);
    }

    private List<InternalRow> readWithFilter(Table table, Predicate filter, int[] projection)
            throws Exception {
        ReadBuilder readBuilder = table.newReadBuilder().withFilter(filter);
        if (projection != null) {
            readBuilder.withProjection(projection);
        }
        RecordReader<InternalRow> reader =
                readBuilder.newRead().createReader(readBuilder.newScan().plan());
        InternalRowSerializer serializer =
                new InternalRowSerializer(
                        projection == null
                                ? table.rowType()
                                : Projection.of(projection).project(table.rowType()));
        List<InternalRow> rows = new ArrayList<>();
        reader.forEachRemaining(row -> rows.add(serializer.copy(row)));
        return rows;
    }
}

File: paimon-python/pypaimon/tests/system/buckets_table_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""End-to-end tests for the ``$buckets`` system table."""

import os
import shutil
import tempfile
import unittest

import pyarrow as pa

from pypaimon import CatalogFactory, Schema
from pypaimon.schema.data_types import DataField
from pypaimon.table.system.buckets_table import BucketsTable


def _read(table):
    rb = table.new_read_builder()
    return rb.new_read().to_arrow(rb.new_scan().plan().splits())


class BucketsTableTest(unittest.TestCase):

    def setUp(self):
        self.tmp = tempfile.mkdtemp(prefix="buckets_sys_")
        warehouse = os.path.join(self.tmp, "warehouse")
        self.catalog = CatalogFactory.create({"warehouse": warehouse})
        self.catalog.create_database("db", False)

    def tearDown(self):
        shutil.rmtree(self.tmp, ignore_errors=True)

    def _create_partitioned_table(self, num_buckets=2):
        fields = [
            DataField.from_dict({"id": 0, "name": "id", "type": "INT"}),
            DataField.from_dict({"id": 1, "name": "v", "type": "STRING"}),
            DataField.from_dict({"id": 2, "name": "dt", "type": "STRING"}),
        ]
        self.catalog.create_table(
            "db.t",
            Schema(
                fields=fields,
                partition_keys=["dt"],
                options={"bucket": str(num_buckets)},
            ),
            False,
        )

    def _write_data(self):
        table = self.catalog.get_table("db.t")
        write_builder = table.new_batch_write_builder()
        writer = write_builder.new_write()
        commit = write_builder.new_commit()
        writer.write_arrow(pa.table({
            "id": pa.array([1, 2, 3, 4], type=pa.int32()),
            "v": ["a", "b", "c", "d"],
            "dt": ["2024-01-01", "2024-01-01", "2024-01-02", "2024-01-02"],
        }))
        commit.commit(writer.prepare_commit())
        writer.close()
        commit.close()

    def test_buckets_table_loaded_via_catalog(self):
        self._create_partitioned_table()
        table = self.catalog.get_table("db.t$buckets")
        self.assertIsInstance(table, BucketsTable)

    def test_schema_column_layout(self):
        self._create_partitioned_table()
        table = self.catalog.get_table("db.t$buckets")
        row_type = table.row_type()
        expected = [
            ("partition", True), ("bucket", False),
            ("record_count", False), ("file_size_in_bytes", False),
            ("file_count", False), ("last_update_time", True),
        ]
        self.assertEqual([n for n, _ in expected],
                         [f.name for f in row_type.fields])
        for field, (_, expected_nullable) in zip(row_type.fields, expected):
            self.assertEqual(expected_nullable, field.type.nullable,
                             "field {} nullability".format(field.name))
        self.assertEqual(["partition", "bucket"], table.primary_keys())

    def test_empty_when_no_snapshot_exists(self):
        self._create_partitioned_table()
        arrow_table = _read(self.catalog.get_table("db.t$buckets"))
        self.assertEqual(0, arrow_table.num_rows)

    def test_aggregates_by_partition_and_bucket(self):
        self._create_partitioned_table(num_buckets=2)
        self._write_data()

        arrow_table = _read(self.catalog.get_table("db.t$buckets"))
        self.assertGreater(arrow_table.num_rows, 0)

        partitions = arrow_table.column("partition").to_pylist()
        self.assertTrue(all(p in ("dt=2024-01-01", "dt=2024-01-02")
                            for p in partitions))

        for size in arrow_table.column("file_size_in_bytes").to_pylist():
            self.assertGreater(size, 0)
        for count in arrow_table.column("file_count").to_pylist():
            self.assertGreaterEqual(count, 1)

        total_records = sum(arrow_table.column("record_count").to_pylist())
        self.assertEqual(4, total_records)

    def test_rows_sorted_by_partition_then_bucket(self):
        self._create_partitioned_table(num_buckets=2)
        self._write_data()

        arrow_table = _read(self.catalog.get_table("db.t$buckets"))
        partitions = arrow_table.column("partition").to_pylist()
        buckets = arrow_table.column("bucket").to_pylist()

        rows = list(zip(partitions, buckets))
        self.assertEqual(rows, sorted(rows))


if __name__ == "__main__":
    unittest.main()
